File: ts/src/run/mini.ts
/** Run mini-SWE-agent in your local environment. Default executable `mini`.
 * Ported from src/minisweagent/run/mini.py */
import path from "node:path";
import os from "node:os";
import chalk from "chalk";
import { globalConfigDir } from "../index.js";
import { getAgent } from "../agents/index.js";
import { getEnvironment } from "../environments/index.js";
import { getModel } from "../models/index.js";
import { builtinConfigDir, getConfigFromSpec } from "../config/index.js";
import { recursiveMerge, UNSET, type MaybeUnset } from "../utils/serialize.js";
import readline from "node:readline";

export const DEFAULT_CONFIG_FILE = process.env.MSWEA_MINI_CONFIG_PATH
  ? path.resolve(process.env.MSWEA_MINI_CONFIG_PATH)
  : path.join(builtinConfigDir, "mini.yaml");
export const DEFAULT_OUTPUT_FILE = path.join(globalConfigDir, "last_mini_run.traj.json");

export interface MiniOptions {
  model?: string;
  modelClass?: string;
  agentClass?: string;
  environmentClass?: string;
  task?: string;
  yolo?: boolean;
  costLimit?: number;
  configSpec?: string[];
  output?: string;
  exitImmediately?: boolean;
}

async function multilinePrompt(): Promise<string> {
  return new Promise((resolve) => {
    const lines: string[] = [];
    const rl = readline.createInterface({ input: process.stdin, output: process.stderr });
    process.stderr.write("(multiline - Ctrl+D or empty line to submit)\n");
    rl.on("line", (line) => {
      if (line === "") {
        rl.close();
        resolve(lines.join("\n"));
      } else {
        lines.push(line);
      }
    });
    rl.on("close", () => resolve(lines.join("\n")));
  });
}

export async function runMini(opts: MiniOptions = {}): Promise<unknown> {
  const configSpec = opts.configSpec ?? [DEFAULT_CONFIG_FILE];
  process.stderr.write(`Building agent config from specs: ${chalk.green(configSpec.join(", "))}\n`);

  const configs = configSpec.map((spec) => getConfigFromSpec(spec));
  configs.push({
    run: { task: opts.task ?? UNSET },
    agent: {
      agent_class: opts.agentClass ?? UNSET,
      mode: opts.yolo ? "yolo" : UNSET,
      cost_limit: opts.costLimit !== undefined ? opts.costLimit : UNSET,
      confirm_exit: opts.exitImmediately ? false : UNSET,
      output_path: opts.output ?? UNSET,
    },
    model: {
      model_class: opts.modelClass ?? UNSET,
      model_name: opts.model ?? UNSET,
    },
    environment: {
      environment_class: opts.environmentClass ?? UNSET,
    },
  });
  const config = recursiveMerge(...configs);

  let runTask = ((config.run as Record<string, unknown>)?.task as MaybeUnset<unknown>) ?? UNSET;
  if (runTask === UNSET) {
    process.stderr.write(chalk.yellow("What do you want to do?\n"));
    runTask = await multilinePrompt();
    process.stderr.write(chalk.green("Got that, thanks!\n"));
  }

  const model = getModel(undefined, (config.model as Record<string, unknown>) ?? {});
  const env = getEnvironment((config.environment as Record<string, unknown>) ?? {}, "local");
  const agent = getAgent(model, env, (config.agent as Record<string, unknown>) ?? {}, "interactive");
  await agent.run(runTask as string);

  const outputPath = (config.agent as Record<string, unknown>)?.output_path as string | undefined;
  if (outputPath) {
    process.stderr.write(`Saved trajectory to ${chalk.green(outputPath)}\n`);
  }
  return agent;
}

File: go/internal/run/mini.go
// Package run provides the mini run script.
// Ported from src/minisweagent/run/mini.py
package run

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/joho/godotenv"
	"github.com/tonythethompson/mini-swe-agent/go/internal/agent"
	"github.com/tonythethompson/mini-swe-agent/go/internal/config"
	"github.com/tonythethompson/mini-swe-agent/go/internal/env"
	"github.com/tonythethompson/mini-swe-agent/go/internal/model"
	"github.com/tonythethompson/mini-swe-agent/go/internal/serialize"
)

// MiniOptions holds CLI options for the mini command.
type MiniOptions struct {
	Model            string
	ModelClass       string
	AgentClass       string
	EnvironmentClass string
	Task             string
	Yolo             bool
	CostLimit        float64
	ConfigSpec       []string
	Output           string
	ExitImmediately  bool
}

// RunMini runs the mini agent.
func RunMini(opts MiniOptions) error {
	// Load .env
	configDir := os.Getenv("MSWEA_GLOBAL_CONFIG_DIR")
	if configDir == "" {
		configDir = filepath.Join(os.Getenv("HOME"), ".config", "mini-swe-agent")
	}
	godotenv.Load(filepath.Join(configDir, ".env"))

	defaultConfigFile := os.Getenv("MSWEA_MINI_CONFIG_PATH")
	if defaultConfigFile == "" {
		defaultConfigFile = filepath.Join(config.BuiltinConfigDir, "mini.yaml")
	}

	configSpecs := opts.ConfigSpec
	if len(configSpecs) == 0 {
		configSpecs = []string{defaultConfigFile}
	}

	fmt.Fprintf(os.Stderr, "Building agent config from specs: %v\n", configSpecs)

	var configs []map[string]any
	for _, spec := range configSpecs {
		c, err := config.GetConfigFromSpec(spec)
		if err != nil {
			return fmt.Errorf("error loading config %s: %w", spec, err)
		}
		configs = append(configs, c)
	}

	// Add CLI overrides
	cliConfig := map[string]any{
		"run": map[string]any{"task": opts.Task},
		"agent": map[string]any{
			"agent_class":  opts.AgentClass,
			"mode":         "",
			"cost_limit":   opts.CostLimit,
			"confirm_exit": !opts.ExitImmediately,
			"output_path":  opts.Output,
		},
		"model": map[string]any{
			"model_class": opts.ModelClass,
			"model_name":  opts.Model,
		},
		"environment": map[string]any{
			"environment_class": opts.EnvironmentClass,
		},
	}
	if opts.Yolo {
		cliConfig["agent"].(map[string]any)["mode"] = "yolo"
	}
	if opts.Task == "" {
		cliConfig["run"].(map[string]any)["task"] = serialize.Unset
	}
	if opts.AgentClass == "" {
		cliConfig["agent"].(map[string]any)["agent_class"] = serialize.Unset
	}
	if opts.ModelClass == "" {
		cliConfig["model"].(map[string]any)["model_class"] = serialize.Unset
	}
	if opts.Model == "" {
		cliConfig["model"].(map[string]any)["model_name"] = serialize.Unset
	}
	if opts.EnvironmentClass == "" {
		cliConfig["environment"].(map[string]any)["environment_class"] = serialize.Unset
	}
	if opts.CostLimit == 0 {
		cliConfig["agent"].(map[string]any)["cost_limit"] = serialize.Unset
	}

	configs = append(configs, cliConfig)
	merged := serialize.RecursiveMerge(configs...)

	// Get task
	runTask := ""
	if runCfg, ok := merged["run"].(map[string]any); ok {
		if t, ok := runCfg["task"].(string); ok {
			runTask = t
		}
	}
	if runTask == "" {
		fmt.Fprintln(os.Stderr, "What do you want to do?")
		// Read from stdin
		line := ""
		fmt.Scanln(&line)
		runTask = line
	}

	// Build model, env, agent
	modelCfg := map[string]any{}
	if m, ok := merged["model"].(map[string]any); ok {
		modelCfg = m
	}
	m, err := model.GetModel(modelCfg)
	if err != nil {
		return err
	}

	envCfg := map[string]any{}
	if e, ok := merged["environment"].(map[string]any); ok {
		envCfg = e
	}
	e, err := env.GetEnvironment(envCfg, "local")
	if err != nil {
		return err
	}

	agentCfg := map[string]any{}
	if a, ok := merged["agent"].(map[string]any); ok {
		agentCfg = a
	}
	a, err := agent.GetAgent(m, e, agentCfg, "interactive")
	if err != nil {
		return err
	}

	_, err = a.Run(runTask)
	if err != nil {
		return err
	}

	if outputPath, ok := agentCfg["output_path"].(string); ok && outputPath != "" {
		fmt.Fprintf(os.Stderr, "Saved trajectory to %s\n", outputPath)
	}
	return nil
}
